### DISCUSSION TITLE
Integrate translations into beta workflow?

### DISCUSSION THREAD

**Discussion by joelnitta** [carpentries/workbench#6] [General]
When the [blog post](https://carpentries.org/blog/2022/05/workbench-beta/#pre-beta) announcing the beta phase of the Workbench format was written, there was no consideration yet as to how to handle translations.

Now that there is a (very early!) [development version of a package 'dovetail'](https://github.com/joelnitta/dovetail) to help with translating lessons in Workbench, could we discuss how this might be integrated into the beta workflow? It may be a bit premature since `dovetail` is still just getting started, but I would at least like to brainstorm how this could be done. 

One reason to think about this sooner rather than later is that we have local communities who would like to contribute translations so lessons may be used in their local languages, but it would be preferable to direct this effort at Workbench lessons (even in beta), since 'style' lessons will be phased out soon.

**Comment by zkamvar** [comment]
Joel, I cannot tell you how impressed I am at what you created! I think testing this out is absolutely needed and worthwhile and we should figure out the logistics so that we don't end up overwhelming the maintainers of the lessons undergoing the beta phase.

I think the most important thing is that we establish a clear workflow of what the translation workflow (inputs and outputs) looks like regardless of the tools used to create it (which I think you did well in dovetail). This way, when we need to in the future, we can pivot to a system that works well to be easy and equitable for translators and maintainers alike. 

I also want to apologise for not getting more infrastructure for translations. Getting the workbench setup has been a long journey and we have had translation on our radar from the beginning, but it's been a big enough challenge that I haven't been able to properly tackle it (that is, make sure the solution is equitable) quite just yet. 

## Chopportunities

There are a few logistical challenges that I'm wondering (and have been wondering) about 

1. what does the development cycle look like for translators and maintainers?
2. Is this language-agnostic? I.E. would this work for any language to any other language (e.g. can we have a Japanese lesson as the origin and translate that into Spanish)?
3. Where would the translations live?
4. Where does the website live?
5. How do we give proper credit the translators?

Regarding point 1 and 3: This is particularly important to me because the model of keeping po files inside the lesson repository means that the number of pull requests to that repository increases with the number of languages added, which puts burdens on the maintainers and barriers for the translators.

Regarding point 4: I think this point can be saved for later down the road, but I do want to get my thoughts in here. In the workbench website templates, we have [a reserved navigation item for language selection next to the instructor view button](https://github.com/carpentries/varnish/blob/e9f5f8764e5be0b422560981db5536633ee3c51b/inst/pkgdown/templates/navbar.html#L43-L61), and [a framework for implementing l10n into the lesson infrastructure](https://github.com/carpentries/sandpaper/issues/205). The challenge is: [where do the translated websites live](https://github.com/carpentries/sandpaper/issues/216)? 

Sorry if this raised more questions than answers.

**Reply by joelnitta** [reply]
> Sorry if this raised more questions than answers

No worries at all! This is exactly what I was hoping for: a thorough discussion about these topics.

> I also want to apologise for not getting more infrastructure for translations. Getting the workbench setup has been a long journey and we have had translation on our radar from the beginning, but it's been a big enough challenge that I haven't been able to properly tackle it (that is, make sure the solution is equitable) quite just yet.

Again no problem, but I think Workbench is now mature enough to start considering this... to put it another way, **I don't want translations to become an afterthought**. I would like us to start considering this now so that design decisions can be made that facilitate translations instead of us having to figure out awkward solutions around an existing setup later.

My thoughts on the points you raised:

1. I am still figuring that out myself. To answer very basically, just considering the case if the MD file changes: I have not done enough testing yet, but re-running `md2po()` on a modified MD file should flag out-of date translations (as ["fuzzy"](https://www.gnu.org/software/gettext/manual/html_node/Fuzzy-Entries.html)), delete translations if their original text is gone, and add new `msgid`/`msgtr` fields for new strings. So the translators would need to update the PO file accordingly, then the translation could be built again. In this sense it's not any different from a standard translation via PO file workflow. (Note that the new `msgid`/`msgtr` fields get added to the end of the PO file, not in their original order in the MD file, which could be confusing for translators. I'm thinking to add a script to re-order the PO file to match the MD file after updating.)

2. Theoretically I think there is no limitation on the language of the input/output MD, but again I need to test.

3. Depends what you mean by "translations" (PO file, translated MD files, or translated rendered HTML lesson). Currently I think the PO files should live with the lesson.

4. My (very rough!) idea is that there could be one branch per rendered, translated lesson, which could then have subdomains. You can see an example of this actually with my website, https://joelnitta.com (source: https://github.com/joelnitta/joelnitta-home), which has `gh-pages` and `gh-pages-ja` branches, which are served at https://www.joelnitta.com and https://www.ja.joelnitta.com/, respectively. This is akin to one of your proposals for [sandpaper #216](https://github.com/carpentries/sandpaper/issues/216). But I'm not sure how these would be implemented with the [language selection navigation](https://github.com/carpentries/varnish/blob/e9f5f8764e5be0b422560981db5536633ee3c51b/inst/pkgdown/templates/navbar.html#L43-L61).

5. My [idea (and strong hope)](https://github.com/carpentries/carpentries.org/issues/435#issuecomment-1114169050) is that translators be recognized the same way as instructors and maintainers. **I want translation to receive equal status as lesson development, maintenance, and teaching.** Much more to discuss about this, that is just a starting point.

> Regarding point 1 and 3: This is particularly important to me because the model of keeping po files inside the lesson repository means that the number of pull requests to that repository increases with the number of languages added, which puts burdens on the maintainers and barriers for the translators.

Yes, but what is the alternative? In our [previous setup with PO files residing in their own repo](https://github.com/carpentries-i18n/i18n), we had to use submodules to track lesson versions, and rendered lessons each lived in their own repo. This necessitated an elaborate dance of repo forking, pulling, pushing, etc, and it was very hard to ever get everything "in sync" (AFAIK has not been achieved yet). Having tried to use that model, I would say the burden was quite high (on the translation team, at least). There may be other options, but IMHO including the PO file in the lesson will be easier. (None of this is to criticize @dpshelio, who did a ton of work to put this system together - I just think we have learned lessons from this experience that we now have a chance to build on with the transition to Workbench).

It is also a bit of a philosophical question: what do you consider the lesson contents? If that includes translations, I think it makes sense to include them with the lesson.

AFAIK, including PO files alongside the source code is standard practice (e.g., [in R packages](https://cran.ism.ac.jp/doc/manuals/r-devel/R-exts.html#Internationalization)).

One way to reduce the number of files would be to use one PO file per language instead of one PO file per MD file per language. But that makes collaborative editing of the PO files difficult (more merge conflicts) because they become quite large (thousands of lines). 

RE: [sandpaper #205](https://github.com/carpentries/sandpaper/issues/205), I did not know about the idea to use R's standard PO system for {pkgdown}. But after taking a glance [at that issue](https://github.com/r-lib/pkgdown/issues/1446#issuecomment-953282411), I think their scope is only [webpage framework](https://github.com/r-lib/pkgdown/issues/1446#issuecomment-946098576), not content. We need to do both. I think [mdpo looks more promising for markdown prose content](https://mdpo.readthedocs.io/en/master/rationale.html) like we have in lessons.

**Reply by zkamvar** [reply]
Hi Joel, I've seen your reply and I will reply soon... I'm just a bit overwhelmed right now. 

In short: I agree that translations should not be an afterthought. I think we have a lot of fodder to draft up a project proposal to move forward in a strategic manner.

**Reply by zkamvar** [reply]
Joel, I want you to know that I had a detailed response with diagrams and everything that ended up getting deleted because I accidentally clicked a link on the preview and all my changes disappeared 😭

**Reply by zkamvar** [reply]
In short: 

 - I agree with you and I think we can do this in a way that avoids the translators needing to know anything about Git, Github, R, Python, Docker, etc. because automation is better
 - My brain works very slowly, so my hesitancy to say "yeah! let's go for it!" is not because I lack enthusiasm or urgency for this issue: it's literally because I'm not wired to do that.
 - I will bring this to the curriculum team next week in addition to being on the call in two weeks. 
 - Because we have the advantage of automation, there are several ways to approach this from the standpoint of where the files live and each one has its own advantages and pitfalls (insert diagrams that Zhian lost forever)

**Reply by joelnitta** [reply]
@zkamvar Thanks for bringing this up with the curriculum team. I am very interested to hear what others think.

I would also be quite interested to see the diagrams sketching out your ideas for how to organize files. I hope they can be recreated without too much effort!

**Reply by dpshelio** [reply]
> including the PO file in the lesson will be easier. 

I agree! The reason the PO files were in a separate repo was to summarise the progress of translations across multiple lessons and to progress without having to add more responsibilities to the maintainers to accept changes they may not understand.

Nowadays, and thanks to GitHub actions, automation can do a lot more! Besides, to keep a track of the translation progress is probably better done on the translations platforms (which it's probably where most of the translators will be). 

The PO files can be updated with a cron job that pulls the bits that have been already reviewed and accepted.

**Reply by joelnitta** [reply]
@dpshelio I'm glad you agree. I think there is a lot of flexibility about where the translated (R)md files and rendered webpages go, but I feel rather strongly about including the PO files in the lesson repos for the lessons I mentioned above (no need for submodules or version tracking across original lesson and PO; it makes it easy to locally render the translation; PO files *are* source).

**Reply by zkamvar** [reply]
I agree that the po files living inside the repository can indeed work, and there are two ways it can work (each with tradeoffs) (and I apologise if I'm being too verbose; I have to process my thoughts out loud in order to understand them):

```mermaid
flowchart LR
M["😼 Maintainer"]
GH[(Repository)]

TF[(Transifex)]
T["😹 Translator"]

GH -.->|automation| TF;
T --> TF;
TF -.->|automation| GH;
GH --> M;
M --> GH;
```

In the above scenario, the translations get pulled in and automatically committed with no need for the maintainer's approval. However, if the maintainers are used to pushing directly to main, they will encounter this scenario:

```
$ git push
To https://github.com/program/lesson.git
 ! [rejected]        main -> main (non-fast-forward)
error: failed to push some refs to 'https://github.com/program/lesson.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

This can be fixed by having maintainers always pull before they push, but it can be disruptive. This is why the [updates to workflows and the package cache are done via pull requests](https://carpentries.github.io/sandpaper/articles/automated-pull-requests.html). A diagram for this for the translation workflow can be seen here:

```mermaid
flowchart LR
M["😼 Maintainer"]
PR>"Pull Request"]
GH[(Repository)]


TF[(Transifex)]
T["😹 Translator"]

GH -.->|automation| TF;
T --> TF;
TF -.->|automation| PR --> GH;
M --> PR;
```

The drawback for this is that a maintainer is required to merge the pull requests as they come in, which means more notifications for the maintainers, which they have expressed frustration with. 


The advantage of hosting the translations in a separate repository is that it could be an r-universe style infrastructure where the translation repository is completely automated in fetching updates from the lessons and the PO files:


```mermaid
flowchart LR
M["😼 Maintainer"]
GH[(Repository)]
AU[("Translation Repo 🤖")]
TF[(Transifex)]
T["😹 Translator"]

M --> GH
GH -.->|automation| AU
AU -.->|automation| TF
TF -.->|automation| AU
T --> TF
```

The benefit is that both the translators and maintainers can work seamlessly and new languages can be easily added. The cost is that it requires a more centralised setup that takes a bit more time to get off the ground and it's more difficult for orgs outside of The Carpentries to integrate this workflow. 


I think either way is possible and I think it's possible to switch between the two.

**Reply by joelnitta** [reply]
I see, thanks for taking the time to re-create the diagrams!

I think it will be hard to make a final call without testing out each of these options. What about at least trying one of them during the beta phase?

From the standpoint of [dovetail](https://github.com/joelnitta/dovetail), I think it could be compatible with any one of these scenarios, so I will continue working on it with these possibilities in mind.

**Reply by zkamvar** [reply]
> I think it will be hard to make a final call without testing out each of these options. What about at least trying one of them during the beta phase?

Joel, I really want to say yes and I really want to get this project going, but the problem is that I literally do not have enough hours in my day to support the additional coordination and development required for this testing to happen in parallel with the beta phase initially without burning out.

**Reply by zkamvar** [reply]
So, it's not a yes, but it's not quite a no... it's a please give me some time because there are a lot of complexities running around.

**Reply by joelnitta** [reply]
No problem! Better to be realistic about expectations than to overdo it. In the meanwhile, I can continue with development of dovetail on my own and see how it works with test translations of the beta lessons in the [Japanese community](https://swcarpentry-ja.github.io/en/).
